cli/internal/globwatcher/globwatcher.go

```diff
@@ -135,24 +146,55 @@ func (g *GlobWatcher) OnFileWatchEvent(ev filewatcher.Event) {
 			g.logger.Error(fmt.Sprintf("failed to check path %v against glob %v: %v", repoRelativePath, glob, err))
 			continue
 		}
-		// If this glob matches, we know that it has changed for every hash that included this glob.
+		// If this glob matches, we know that it has changed for every hash that included this glob
+		// and is not excluded by a hash's exclusion globs.
 		// So, we can delete this glob from every hash tracking it as well as stop watching this glob.
 		// To stop watching, we unref each of the directories corresponding to this glob.
 		if matches {
-			delete(g.globStatus, glob)
+			// We need to check if all the hashes exclude this path. In which case we shouldn't remove it
+			// from globStatus
+			allExcluded := true
+
 			for hashUntyped := range hashStatus {
 				hash := hashUntyped.(string)
 				hashGlobs, ok := g.hashGlobs[hash]
+				isExcluded := false
```

Comment:
You need to check `ok` before using `hashGlobs`. If `ok` is false, you aren't guaranteed that `hashGlobs` is non-nil. The check lower down should probably be moved up.